cleverhans/attack_bundling.py

```diff
@@ -0,0 +1,797 @@
+  pgd_params = copy.copy(threat_params)
+  pgd_params["eps_iter"] = eps_iter
+  pgd_params["nb_iter"] = nb_iter
+  assert batch_size % num_devices == 0
+  dev_batch_size = batch_size // num_devices
+  ones = tf.ones(dev_batch_size, tf.int32)
+  expensive_pgd = []
+  for cls in range(nb_classes):
+    cls_params = copy.copy(pgd_params)
+    cls_params['y_target'] = tf.to_float(tf.one_hot(ones * cls, nb_classes))
+    cls_attack_config = AttackConfig(pgd_attack, cls_params, "pgd_" + str(cls))
+    pgd_attack_configs.append(cls_attack_config)
+    expensive_params = copy.copy(cls_params)
+    expensive_params["eps_iter"] /= 25.
+    expensive_params["nb_iter"] *= 25.
+    expensive_config = AttackConfig(pgd_attack, expensive_params, "expensive_pgd_" + str(cls))
+    expensive_pgd.append(expensive_config)
+  attack_configs = [noise_attack_config] + pgd_attack_configs + expensive_pgd
+  new_work_goal = {config: 5 for config in attack_configs}
+  pgd_work_goal = {config: 5 for config in pgd_attack_configs}
+  # TODO: lower priority: make sure bundler won't waste time running targeted
+  # attacks on examples where the target class is the true class
+  goals = [Misclassify(new_work_goal={noise_attack_config: 50}),
+           Misclassify(new_work_goal=pgd_work_goal),
+           MaxConfidence(t=0.5, new_work_goal=new_work_goal),
+           MaxConfidence(t=0.75, new_work_goal=new_work_goal),
+           MaxConfidence(t=0.875, new_work_goal=new_work_goal),
+           MaxConfidence(t=0.9375, new_work_goal=new_work_goal),
+           MaxConfidence(t=0.96875, new_work_goal=new_work_goal),
+           MaxConfidence(t=0.984375, new_work_goal=new_work_goal),
+           MaxConfidence(t=1.)]
+  bundle_attacks(sess, model, x, y, attack_configs, goals, report_path)
+  # This runs forever
+
+class AttackConfig(object):
+  """
+  An attack and associated parameters.
+  :param attack: cleverhans.attacks.Attack
+  :param params: dict of keyword arguments to pass to attack.generate
+  :param name: str, name to be returned by __str__ / __repr__
+  """
+  def __init__(self, attack, params=None, name=None):
+    self.attack = attack
+    self.params = params
+    self.name = name
+    if params is not None:
+      assert isinstance(params, dict)
+      for key in params:
+        assert isinstance(key, six.string_types), type(key)
+
+  def __str__(self):
+    if self.name is not None:
+      return self.name
+    return "AttackConfig(" + str(self.attack) + ", " + str(self.params) + ")"
+
+  def __repr__(self):
+    return self.__str__()
+
+def bundle_attacks(sess, model, x, y, attack_configs, goals, report_path):
+  """
+  Runs attack bundling.
+  Users of cleverhans may call this function but are more likely to call
+  one of the recipes above.
+
+  Reference: https://openreview.net/forum?id=H1g0piA9tQ
+
+  :param sess: tf.session.Session
+  :param model: cleverhans.model.Model
+  :param x: numpy array containing clean example inputs to attack
```

Comment:
Given that we have to create the placeholders within the train and evaluate functions of the library to support multi-GPU, I agree it makes sense to reflect that change in the new API. I don't understand though why not keep a name (e.g., x_train and y_train) that makes it clear to users that they should pass numpy arrays and not tensors? If I remember correctly, I think that's what train and batch_eval do in their multi GPU variants. I'm sorry I must have missed this when merging `correctness_and_confidence`. I'm just worried that users will get confused if they miss the change, and having this name convention there would not change the implementation, it just makes it easier for users to notice the change? Maybe I'm missing something?

Regarding the second part of your comment, I'm not sure we want to completely remove the ability to use placeholders directly with the attack classes. I feel like it would hurt flexibility. For the rest of the library (training, eval), I understand why that makes sense and agree we should keep going in that direction. We should probably have this discussion with tf.eager in mind too since it will likely play in the design as well.

Reply:
We don't want to use a name like `x_train` here because this could also be the validation or test set. Since this code is used for benchmarking, it's usually going to be the test set.

A lot of our other naming conventions haven't worked out very well. We've had a lot of "_ph" names for variables that don't actually need to be placeholders. We're also not very consistent with "_val" and "_value". Sometimes "_value" means that a variable is a tensorflow tensor rather than a callable, sometimes "_value" means that a variable is numpy rather than tensorflow. So I find the names are usually just no typing but don't add any information, and sometimes mislead me.

How about using `x` for the name and, in addition to the docstring saying `x` is numpy, also have an if statement that raises a ValueError immediately if `x` is not numpy?

It's still possible for users to make placeholders, but pretty soon the default will be for them to not do that. Also, as time goes on, it will be an increasingly dumb thing for them to do. Already, making your own placeholder is a pretty bad decision, because it means everything you write after that will need to be rewritten if you ever want to upgrade to using the multi-GPU features. So while we still want to allow it, we should give people strong nudges away from doing it.